concordium-consensus/src/Concordium/KonsensusV1/Transactions.hs

```diff
@@ -186,50 +201,57 @@ processBlockItems ::
   where
     -- Create a context suitable for verifying a transaction within a 'Block' context.
     getCtx = do
-        _ctxSkovData <- get
         let _ctxBs = bpState parentPointer
+        -- We base the block energy limit on the chain parameters for the parent block.
+        -- This might not be accurate, because the parameter can change, but exceeding the limit
+        -- is treated as 'MaybeOK', so we will not fail even if the check is incorrect.
+        -- When the block is actually executed, the actual energy limit will be enforced.
         chainParams <- Concordium.GlobalState.BlockState.getChainParameters _ctxBs
         let _ctxMaxBlockEnergy = chainParams ^. cpConsensusParameters . cpBlockEnergyLimit
-        return $! Context{_ctxTransactionOrigin = TVer.Block, ..}
+        return Context{_ctxTransactionOrigin = TVer.Block, ..}
     theRound = bbRound bb
     theTime = bbTimestamp bb
-    -- Process the list of transactions recursively.
-    process :: [BlockItem] -> m Bool
-    process [] = return True
-    process (bi : bis) = do
+    -- Process a transaction
+    process ::
+        Context (PBS.HashedPersistentBlockState pv) ->
+        BlockItem ->
+        MaybeT m (BlockItem, TVer.VerificationResult)
+    process verificationContext bi = do
         let txHash = getHash bi
         tt' <- gets' _transactionTable
         -- Check whether we already have the transaction.
         case tt' ^. TT.ttHashMap . at' txHash of
-            Just (_, results) -> do
+            Just (bi', results) -> do
                 -- If we have received the transaction before we update the maximum committed round
                 -- if the new round is higher.
                 when (TT.commitPoint theRound > results ^. TT.tsCommitPoint) $
                     transactionTable . TT.ttHashMap . at' txHash . mapped . _2 %=! TT.updateCommitPoint theRound
-                -- And we continue processing the remaining transactions.
-                process bis
+                return (bi', results ^. TT.tsVerRes)
             Nothing -> do
                 -- We verify the transaction and check whether it's acceptable i.e. Ok or MaybeOk.
                 -- If that is the case then we add it to the transaction table and pending transactions.
                 -- If it is NotOk then we stop verifying the transactions as the block can never be valid now.
-                !verRes <- runTransactionVerifierT (TVer.verify theTime bi) =<< getCtx
+                !verRes <- verifyBlockItem theTime bi verificationContext
                 case verRes of
                     -- The transaction was deemed non verifiable i.e., it can never be
-                    -- valid. We short circuit the recursion here and return 'False'.
-                    (TVer.NotOk _) -> return False
+                    -- valid. We short circuit the recursion here and return 'Nothing'.
+                    (TVer.NotOk _) -> empty
```

Comment:
Can we please just return Nothing here. This `empty` is needlessly general here.